Framework/Geometry/inc/MantidGeometry/Objects/MeshObject.h

```diff
@@ -61,10 +62,10 @@ Code Documentation is available at: <http://doxygen.mantidproject.org>
 */
 class MANTID_GEOMETRY_DLL MeshObject : public IObject {
 public:
-  /// Default constructor
-  MeshObject();
-  /// Constructor from string.
-  MeshObject(const std::string &shapeXML);
+  /// Main constructor
+  MeshObject(const std::vector<size_t> &faces,
+             const std::vector<Mantid::Kernel::V3D> &vertices);
+
   /// Copy constructor
   MeshObject(const MeshObject &);
```

Comment:
Where is copy and copy-assignment actually required? I believe that in most cases, `IObject` will be wrapped in a `shared_pointer` (in the production code), and those objects would be immutable. So i'm not sure why we would need the ability to copy. Maybe you could investigate why and comment back. I guess you could delete the method and see what the compiler says when you try to compile mantid.

Reply:
MeshObject is not quite immutable. It material may be changed. Also it is good practice to have a copy constructor so we can control copying rather than leave it to a compiler generated function, likewise with default constructor.

Reply:
You are right, there are complex rules governing when a copy or copy-assignment operator would be provisioned automatically. However, it would be best to disable the copy and copy-assignment operators, not just fail to provide an implementation. i.e.

```cpp
MeshObject(const MeshObject& other) = delete;
MeshObject& operator=(const MeshObject& other) = delete;
``` 

Since you have already provided a non-default constructor the compiler will NOT generate a default constructor for you can safely delete the declaration of the default constructor. I see no reason to allow anyone to create a default-constructed `MeshObject` that has no vertex entries.

Does the material really need to be changed after the lifetime of the `MeshObject` has started? i.e. is a setter for the Material type really necessary? Can this be done via a constructor argument instead?

Reply:
The material does need to be changed after construction, because it is a requirement of `IObject`.

Reply:
From a quick search of the code base a pattern I'm seeing a lot is of the form:

```cpp
    auto shape = boost::shared_ptr<IObject>(in_WS->sample().getShape().clone());
    shape->setMaterial(Material("SetInMultipleScattering", neutron, coeff2));
```

i.e. clients of this part of the code are able to provide the material at the same time that the lifetime of the `IObject` derivative starts. In other words, the `IObject` interface may be wrong and you should not be afraid to change it. For example, the above may be better as:

```cpp
    const auto shape = boost::shared_ptr<IObject>(in_WS->sample().getShape().cloneWithMaterial(Material("SetInMultipleScattering", neutron, coeff2)));
```

Notice how the new shape is now also immutable

Reply:
When I investigated this I found cases where an `IObject_sptr` returned by `ComponentCreationHelper` has its material set. Perhaps I could change this `IObject_sptr` back to a shared pointer to `CGSObject` having got rid of most of them.

Reply:
Can you point me to where in `ComponentCreationHelper`?

Reply:
Also I've found this done to examples returned by `shapeFactory::createShape`.

Reply:
I found the following functions that return an `IObject_sptr` that subsequently gets its material set.
```
ComponentCreationHelper::createSphere
ComponentCreationHelper::createCappedCylinder
ShapeFactory().createShape with at least 4 argument lists
Container::createSampleShape

```

Reply:
Yes there are quite a few. The point is that you don't have to have a setter in order to achieve this. As I said above. Adding a method to `IObject` such as:

```cpp
std::unique_ptr<IObject> cloneWithMaterial(const Material& material) const;
```

Would allow you to remove the `setMaterial` method.

Reply:
How would this work for the functions I listed, when there is no cloning.

Reply:
A simplier approach for many of the cases you listed would simply be to pass the the `Material` as an additional argument to `createSphere`, `createCappedCylinder`, that way you can provide it as part of the constructor to shape, or via the public `setMaterial` method on the derived object at the point that you make it (if you have to).

That means that you are not calling the setter on the `IObject` interface. If this is still causing confusion I can pass by the office tomorrow and talk you through it.

Reply:
I thought it might be something like this.

Reply:
We also need to get rid of `setName` from `IObject.h` .

Reply:
and also `setID`, `setVtkGeometryCacheReader` and `setVtkGeometryCacheWriter`. Every remaining function in `IObject.h`  ends with `const =0`.

Reply:
Another thing to investigate is why 
```
virtual boost::shared_ptr<GeometryHandler> getGeometryHandler() = 0;
```
has no `const` before the ` = 0;`.